src/class/net/net_device.c

```diff
@@ -226,33 +249,95 @@ static void ecm_report(bool nc)
 // return false to stall control endpoint (e.g unsupported request)
 bool netd_control_request(uint8_t rhport, tusb_control_request_t const * request)
 {
-  // Handle class request only
-  TU_VERIFY(request->bmRequestType_bit.type == TUSB_REQ_TYPE_CLASS);
+  switch ( request->bmRequestType_bit.type )
+  {
+    case TUSB_REQ_TYPE_STANDARD:
+      switch ( request->bRequest )
+      {
+        case TUSB_REQ_GET_INTERFACE:
+        {
+          uint8_t const req_itfnum = (uint8_t) request->wIndex;
+          TU_VERIFY(_netd_itf.itf_num+1 == req_itfnum);
+
+          tud_control_xfer(rhport, request, &_netd_itf.itf_data_alt, 1);
+        }
+        break;
 
-  TU_VERIFY (_netd_itf.itf_num == request->wIndex);
+        case TUSB_REQ_SET_INTERFACE:
+        {
+          uint8_t const req_itfnum = (uint8_t) request->wIndex;
+          uint8_t const req_alt    = (uint8_t) request->wValue;
+
+          // Only valid for Data Interface with Alternate is either 0 or 1
+          TU_VERIFY(_netd_itf.itf_num+1 == req_itfnum && req_alt < 2);
+
+          // ACM-ECM only: qequest to enable/disable network activities
+          TU_VERIFY(_netd_itf.ecm_mode);
+
+          _netd_itf.itf_data_alt = req_alt;
+
+          if ( _netd_itf.itf_data_alt )
+          {
+            // TODO since we don't actually close endpoint
+            // hack here to not re-open it
+            if ( _netd_itf.ep_in == 0 && _netd_itf.ep_out == 0 )
+            {
+              TU_ASSERT(_netd_itf.ecm_desc_epdata);
+              TU_ASSERT( usbd_open_edpt_pair(rhport, _netd_itf.ecm_desc_epdata, 2, TUSB_XFER_BULK, &_netd_itf.ep_out, &_netd_itf.ep_in) );
+
+              // TODO should be merge with RNDIS's after endpoint opened
+              // Also should have opposite callback for application to disable network !!
+              tud_network_init_cb();
+              can_xmit = true; // we are ready to transmit a packet
+              tud_network_recv_renew(); // prepare for incoming packets
+            }
```

Comment:
Perhaps another hack could be to do a "stall" and then an "unstall" on the two bulk EP, in order to reset their DTOGs, in the case that the EP are already open. But, not necessary, since it's hacky anyway.

Reply:
that is a good way, though let's just leave it this way, to make sure we got an actual error and have motivation to fix it :)